### PR TITLE
Improve stability for mpmath KBN algorithm in poisson_confidence_interval

### LINKED ISSUE #9596: Issues in poisson_conf_interval with kraft-burrows-nousek
### Description

When I use poisson_conf_interval with kraft-burrows-nousek interval, there is something wrong with it. This is the code:
```
>>> from astropy.stats import poisson_conf_interval
>>> poisson_conf_interval(1000,'kraft-burrows-nousek',1,900,0.9)
```
The output is:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/astropy/stats/funcs.py", line 718, in poisson_conf_interval
    cache=True)(n, background, conflevel)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 2091, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 2161, in _vectorize_call
    ufunc, otypes = self._get_ufunc_and_otypes(func=func, args=args)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 2121, in _get_ufunc_and_otypes
    outputs = func(*inputs)
  File "/usr/lib/python3/dist-packages/astropy/stats/funcs.py", line 1244, in _kraft_burrows_nousek
    return _mpmath_kraft_burrows_nousek(N, B, CL)
  File "/usr/lib/python3/dist-packages/astropy/stats/funcs.py", line 1150, in _mpmath_kraft_burrows_nousek
    N = mpf(N)
  File "/usr/lib/python3/dist-packages/mpmath/ctx_mp_python.py", line 77, in __new__
    v._mpf_ = mpf_pos(cls.mpf_convert_arg(val, prec, rounding), prec, rounding)
  File "/usr/lib/python3/dist-packages/mpmath/ctx_mp_python.py", line 96, in mpf_convert_arg
    raise TypeError("cannot create mpf from " + repr(x))
TypeError: cannot create mpf from 1000
```
I check the codes and find that mpf can't convert type numpy.int64, so I modify the code in function `_mpmath_kraft_burrows_nousek(N, B, CL)` in astropy/stats/funcs.py:
```
N = mpf(N)
B = mpf(B)
```
to
```
N = mpf(int(N))
B = mpf(float(B))
```
After that, the output is changed to:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/astropy/stats/funcs.py", line 718, in poisson_conf_interval
    cache=True)(n, background, conflevel)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 2091, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 2161, in _vectorize_call
    ufunc, otypes = self._get_ufunc_and_otypes(func=func, args=args)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 2121, in _get_ufunc_and_otypes
    outputs = func(*inputs)
  File "/usr/lib/python3/dist-packages/astropy/stats/funcs.py", line 1244, in _kraft_burrows_nousek
    return _mpmath_kraft_burrows_nousek(N, B, CL)
  File "/usr/lib/python3/dist-packages/astropy/stats/funcs.py", line 1192, in _mpmath_kraft_burrows_nousek
    S_max = findroot(func, N - B, tol=1e-4)
  File "/usr/lib/python3/dist-packages/mpmath/calculus/optimization.py", line 979, in findroot
    % (norm(f(*xl))**2, tol))
ValueError: Could not find root within given tolerance. (0.0100000000000009105486 > 0.0001)
Try another starting point or tweak arguments.
```
If run with:
`poisson_conf_interval(1000,'kraft-burrows-nousek',1,901,0.9)`
it outputs the result:
```
array([[ 47.95246273],
       [151.84619844]])
```
but with <=900 background, it doesn't continue.
It seems that `S_max = findroot(func, N - B, tol=1e-4)` in `_mpmath_kraft_burrows_nousek(N, B, CL)` can't find the root. If it is modified to `S_max = findroot(func, 155, tol=1e-4)`, the code with 900 background will run correct. It outputs:
```
array([[ 48.94414313],
       [152.85511478]])
```
So, I think this part of code should be modified to imporve it.

### System Details
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0]
Numpy 1.16.2
Scipy 1.1.0
astropy 3.1.2

### PR DESCRIPTION
**Pull request by hamogu**
### Description

The KBN implementation in astropy differs from the suggestion in the KBN
paper in an important way. The paper suggests to integrate, starting
from the most likely value, into both directions. That is conceptually
easy, save, and could be fast. However, it is a pain to implement because
you have to worry about the integration step size, what do do when you
overstep, which parameter (lower or upper bound) needs to be moved next
etc.
So, instead, the astropy implementation reformulates this a root-finding
problem. That has the advantage that out of-the-box root finding algorithms
can be used to take care of steps and tolerance checking, but it opens up a 
of numerical problems that creep in, in particular when the background is 0.
    
So, this PR reworks the algorithm to use safer interval-based root-finding
algorithms. The PR may slow down the code, but it makes it work for all case
I tested (see added tests).
I tried the following grid (but decided not to put all that in the tests
because it runs over a minute):
    
    ```
    import numpy as np
    s = np.array([0., 1, 2, 3, 5, 10, 49, 50, 99, 100, 500, 999, 1000.])
    b = np.array([0, .1, 1, 2, 3, 4.9, 5, 5.1, 10, 49, 50, 99, 100, 500, 999, 99
    cl = np.array([.9, .95, .99])
    
    for i in s:
        for j in b:
            for c in cl:
                try:
                    _mpmath_kraft_burrows_nousek(i, j, c)
                except:
                    print(i, j, c)
    ```
    
fixes #9596

### CODE REVIEW

**Comment by pllim** [comment]
Failure might be related. I'll let subpackage maintainer decide if this is a bug or enhancement (i.e., does it need to be backported to LTS?) and whether `tol` can be adjustable.

**Comment by larrybradley** [comment]
CI test failure seems real.

I would categorize this as a bugfix, so 4.0.2 looks good to me.

**Comment by hamogu** [comment]
Failure is indeed real, but fortunately, it's just a typo in the test. (I slipped the table and put in the number for 99% confidence level, but asked for the 90% conf level in the test).

**Comment by hamogu** [comment]
About tol: I thought about that: It was hardcoded before, mostly to make the code work when the root finding is not doing well. (I'm the original author of this code, reaching back 5 years ago.)
However, there is some scientific reason to believe that higher precision is never needed: Poisson statistics are about integer counts in a low-count regime. When your measurement is quantized at the integer level, a tolerance of 1e-4 is more than enough.

**Comment by hamogu** [comment]
Codevoc complains about three files that I definitely did not touch. This test is new to me (I'm not a regular contributor to astropy), so please advice me on how to proceed.

**Comment by pllim** [comment]
You can probably ignore the codecov failure.

**Comment by hamogu** [comment]
@larrybradley Maybe you have a chance to look at this again? Except for the codecov failure (which is spurious) it all passes.

**Comment by hamogu** [comment]
Ping! @larrybradley Maybe you have a chance to look at this again? Except for the codecov failure (which is spurious) it all passes.

File: astropy/stats/funcs.py
Comment (larrybradley) [inline astropy/stats/funcs.py:1156]: Minor nit since this is a private function, but why did you remove this blank lines?  There should be blanks lines before each section.

File: astropy/stats/funcs.py
Comment (larrybradley) [inline astropy/stats/funcs.py:1166]: here too

File: astropy/stats/funcs.py
Comment (larrybradley) [inline astropy/stats/funcs.py:1170]: and here

**Review by larrybradley — COMMENTED** [review]
Thanks, @hamogu.  I'm not an expert on this algorithm, but the code looks good to me.  Just some minor comments about removed spaces in a docstring.

**Review by larrybradley — APPROVED** [review]

**Comment by hamogu** [comment]
@larrybradley I certainly did not remove those linebreaks on purpose. No idea how they got lost. Anyway, I put them back in. As the original author of this code, I'd like to claim that I'm the expert, but of course I can't review my own PR ;-)

Moritz

**Pull request merged**
